### ISSUE TITLE
Thread-safe implementation

### ISSUE THREAD

**Issue by robfromboulder** [pmem/pmemkv#26]
Lock eliding using TSX?  (assuming that key collisions will be rare)
Background maintenance thread?
Lots of new tests obviously

**Comment by GBuella** [comment]
This task seems to be fun, so I started coding some ideas for it today.

My current idea (might work, or might not work at all) is based the only sensible idea I found on the internet -- creating patches for the tree.

Every time a thread modifies the tree, it creates a patch, i.e. creating new nodes with the appropriate modifications, instead of rearranging fields in existing nodes. This patch is applied to the original tree by setting the appropriate pointers in the `children` arrays in inner nodes.

```
thread#0
1) creates new modified subtree
2) memory barrier
3) parent->children[i] = [pointer to new subtree]
```
Of course the new subtree can contain nodes from the original subtree, only those nodes which would actually be modified need to be reallocated.

```
thread#1  -- doing lookup
1) reads subtree pointer from parent->children[i]
2) memory barrier
3) continues lookup in the subtree
```

Two issues to be resolved:
 * The old subtrees must be kept around, and garbage collected when it it is known that no one accesses them anymore (how can we know that?)
 * In this simple form, only a single thread can make modifications. It needs to be done while holding a mutex (which do lookup routines don't have to hold). I have an idea for that as well, that might complicate things a bit: a modification is first added to some sort of queue (which can be modified by multiple threads) in the tree. Once it is in the queue, the code does a try_lock on the above mentioned mutex for modifications. Whoever manages to acquire the lock with the try_lock call, actually applies the pending modifications to the tree. Also, tree lookup must consider this queue of pending modifications.

Let me know if you think this is stupid, or viable. Cheers.

**Comment by RobDickinson** [comment]
@GBuella, I beg your forgiveness, I thought I had responded to this thread, but I never did.  I'm really curious to contrast this approach with a b-epsilon tree, which is more disruptive to the structure of the tree, but has some other nice qualities.  I've been heads-down on performance testing and tuning lately, but I'd love to get back to this thread in the next few weeks.  I love the idea of having that debate here on GitHub, so that everybody can see how we arrived at the final thread-safe version.  Anyway, sorry it to so long to thank you for the creative energy here.
